fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/mixin/event/lifecycle/MinecraftServerMixin.java

```diff
@@ -73,4 +89,33 @@ private void closeWorld(CallbackInfo ci, Iterator<ServerWorld> worlds, ServerWor
 			ServerBlockEntityEvents.BLOCK_ENTITY_UNLOAD.invoker().onUnload(blockEntity, serverWorld);
 		}
 	}
+
+	// The locals you have to manage for an inject are insane. And do it twice. A redirect is much cleaner.
+	// Here is what it looks like with an inject: https://gist.github.com/i509VCB/f80077cc536eb4dba62b794eba5611c1
+	@Redirect(method = "createWorlds", at = @At(value = "INVOKE", target = "Ljava/util/Map;put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;"))
+	private <K, V> V onLoadWorld(Map<K, V> worlds, K registryKey, V serverWorld) {
+		final V result = worlds.put(registryKey, serverWorld);
+		ServerWorldEvents.LOAD.invoker().onWorldLoad((MinecraftServer) (Object) this, (ServerWorld) serverWorld);
+
+		return result;
+	}
+
+	@Inject(method = "reloadResources", at = @At("HEAD"))
+	private void beforeResourceReload(Collection<String> collection, CallbackInfoReturnable<CompletableFuture<Void>> cir) {
+		ServerLifecycleEvents.START_DATA_PACK_RELOAD.invoker().startDataPackReload((MinecraftServer) (Object) this, this.serverResourceManager);
+	}
+
+	@Inject(method = "method_29440(Ljava/util/Collection;Lnet/minecraft/resource/ServerResourceManager;)V", at = @At("TAIL"))
+	private void afterResourceReload(Collection<String> enabledPacks, ServerResourceManager serverResourceManager, CallbackInfo ci) {
+		ServerLifecycleEvents.END_DATA_PACK_RELOAD.invoker().endDataPackReload((MinecraftServer) (Object) this, this.serverResourceManager, true);
+	}
+
+	@Inject(method = "reloadResources", at = @At("TAIL"))
+	private void addResourceReloadFailureCallback(Collection<String> collection, CallbackInfoReturnable<CompletableFuture<Void>> cir) {
+		// Hook into fail
+		cir.getReturnValue().exceptionally(throwable -> {
```

Comment:
I highly doubt the function passed to `exceptionally` runs on completable future's default executor's thread than on the server thread.

Reply:
I actually tested that, it does run on server thread.

Reply:
> I actually tested that

This sure runs on server thread if the completable future is completed. but if it isn't completed at that time, think something wrong will happen. Java only added a safe version of exceptionally `exceptionallyAsync` in 12
https://docs.oracle.com/en/java/javase/12/docs/api/java.base/java/util/concurrent/CompletionStage.html#exceptionallyAsync(java.util.function.Function,java.util.concurrent.Executor)

As a result you should use `handleAsync` instead https://docs.oracle.com/en/java/javase/12/docs/api/java.base/java/util/concurrent/CompletionStage.html#handleAsync(java.util.function.BiFunction,java.util.concurrent.Executor)

see https://stackoverflow.com/questions/38254268/completablefuture-exceptionally-with-executor

Reply:
So handleAsync actually reduces the amount of injects we need by one. Should this be using the future's current thread or the server thread?

Reply:
Handle async with server executor would probably be executed 1 tick after the reload is done given how server task scheduling works, but should be safe if we look at data packs' functions ("minecraft:load" function tag is executed 1 tick late too) 👍 so your current work should be great